### ISSUE TITLE
mongodb: cannot plan query with grouping and function of three args

### ISSUE THREAD

**Issue by mossprescott** [slamdata/slamengine#521]
When a function of three args (e.g. `substring`, `cond`, `between`) is used with all non-literal arguments, and grouping is happening, there are multiple problems:
1. Sources won't merge.
2. We're building DocBuilders on DocBuilders, which is not handled in `objectConcat`. These should be inlined, if we can't make `expr3` smarter and just not generate them.

See the pending test (as of #515) with this query:
```sql
select case when pop > 1000 then city else lower(city) end, count(*) from zips group by city
```

**Comment by mossprescott** [comment]
After making another attempt at this, it seems like any kind of general solution will involve handling a number of possible shapes for each of the three arguments--a kind of code we already have too much of.

@sellout and I talked it over and it seems like it would be wise to put this off until we have a better way of keeping track of the "implicit" grouping that makes this problematic. See #562.

**Comment by sellout** [comment]
I have this fixed in a stash that requires PR #574. It uses `fold1Builders` rather than `expr3`’s `nest`. I think with a few changes, `expr1` and `expr2` could use this same function, so we’d have a single `expr(List[WorkflowBuilder])(List[ExprOp] => ExprOp)` operation.
